service/storage_service.cc

```diff
@@ -2658,20 +2658,207 @@ void storage_service::add_expire_time_if_found(inet_address endpoint, int64_t ex
     }
 }
 
-// For more details, see the commends on column_family::load_new_sstables
-// All the global operations are going to happen here, and just the reloading happens
-// in there.
-future<> storage_service::load_new_sstables(sstring ks_name, sstring cf_name) {
-    if (_loading_new_sstables) {
-        throw std::runtime_error("Already loading SSTables. Try again later");
-    } else {
-        _loading_new_sstables = true;
+class send_meta_data {
+    gms::inet_address _node;
+    rpc::sink<frozen_mutation_fragment, streaming::stream_mutation_fragments_cmd> _sink;
+    rpc::source<int32_t> _source;
+    bool _error_from_peer = false;
+    size_t _num_partitions_sent = 0;
+    size_t _num_bytes_sent = 0;
+public:
+    send_meta_data(gms::inet_address node,
+            rpc::sink<frozen_mutation_fragment, streaming::stream_mutation_fragments_cmd> sink,
+            rpc::source<int32_t> source)
+        : _node(std::move(node))
+        , _sink(std::move(sink))
+        , _source(std::move(source)) {
+    }
+    void receive() {
+        for (;;) {
+            std::optional<std::tuple<int32_t>> status_opt = _source().get0();
+            if (!status_opt) {
+                slogger.debug("Finished reading source from node={}", _node);
+                break;
+            }
+            auto status = std::get<0>(*status_opt);
+            slogger.debug("Got error code from node={}, status={}", _node, status);
+            if (status == -1) {
+                _error_from_peer = true;
```

Comment:
This bad error handling in rpc stream is really sad.

Reply:
Yes, I agree.